draft-barnes-acme.md

```diff
@@ -218,304 +219,275 @@ In this section, we describe the four certificate management functions that ACME
-    [0, 2], [1, 2]
+    "recoveryToken": {}
+  },
+
+  "combinations": [ 
+    ["simpleHttps", "recoveryToken"],
+    ["dns", "recoveryToken"]
   ]
 }
+
 ~~~~~~~~~~
 
-In order to avoid replay attacks, the server MUST generate a fresh nonce of at least 128 bits for each authorization transaction, and MUST NOT accept more than one authorizationRequest with the same nonce.
+The client needs to respond with information to complete the challenges.  To do this, the client updates the authorization object received from the server by filling in any required information in the elements of the "challenges" dictionary.  For example, if the client wishes to complete the "simpleHttps" challenge, it needs to provide the "path" component.  (This is also the stage where the client should perform any actions required by the challenge.)
 
-<!-- NOTE: Should we allow multiple keys to sign over the same nonce?  Could be handy for authorizing multiple keys, but seems to also have replay risk.  This could also be addressed by having the "challengeRequest" contain the public key(s). -->
+The client sends these updates back to the server in the form of a JSON merge patch {{RFC7386}} to the authorization document, carried in a POST request to the authorization URI (not the new-authorization URI).  Using a patch allows the client to send information only for challenges it is responding to.  (It is also harmless to send the whole authorization document.)
 
-The client SHOULD satisfy all challenges in one of the sets expressed in the "combinations" array.  If a "combinations" field is not specified, the client SHOULD attempt to fulfill as many challenges as possible.
+~~~~~~~~~~
 
-Once the client believes that it has fulfilled enough challenges, it creates an authorizationRequest object requesting authorization of a key pair for this identifier based on its responses.  The authorizationRequest also contains the public key to be authorized, and the signature by the corresponding private key over the nonce in the challenge.
+POST /acme/authz/asdf HTTP/1.1
+Host: example.com
 
-type (required, string):
-: "authorizationRequest"
+{
+  "challenges": {
+    "simpleHttps": {
+      "path": "Hf5GrX4Q7EBax9hc2jJnfw"
+    },
+    "recoveryToken": {
+      "token": "23029d88d9e123e"
+    }
+  }
+}
+/* Signed as JWS */
 
-sessionID (required, string):
-: The session ID provided by the server in the challenge message (to allow the server to correlate the two transactions).
+~~~~~~~~~~
 
-nonce (required, string):
-: The nonce provided by the server in the challenge message.
+The server updates the authorization document by applying the patch.  Before applying the patch, however, the server MUST delete any fields in the patch besides "challenges" and "contact", since these are the only fields that the client may modify.  Similar constraints may need to be applied to individual challenges (e.g., not letting the client modify the "token" field in a "simpleHttps" challenge).
```

Comment:
This approach invites implementor errors that would be catastrophic, like allowing overwrite of token fields. With such a narrow set of allowed fields from the client, why use JSON merge patch at all? I am much more comfortable defining a ChallengeResponse as its own type of object with explicit handling, as the previous version of this document did.